lib/component.js

```diff
@@ -302,11 +332,14 @@ class Component extends WebComponent {
     }
   }
 
-  updateSelfAndChildren(state) {
-    if (this.initialized && this.shouldUpdate(state)) {
-      this.domPatcher.update(state);
+  updateSelfAndChildren(stateUpdate, options={}) {
+    if (this.initialized && this.shouldUpdate(stateUpdate)) {
+      const newState = Object.assign(this.state, stateUpdate);
+      this.domPatcher.update(newState);
       for (let child of this.$panelChildren) {
-        child.updateSelfAndChildren(state);
+        if (options.exclude !== child) {
+          child.updateSelfAndChildren(child.sharedStateEntries(newState), options);
```

Comment:
Yeah, that's how the current implementation works. Doesn't have to be that way though, either one is a similar implementation - conceptually the difference is whether we want state-sharing to be tree-like or wheel-like (every child having a direct relation to the central app store rather than filtered through the component hierarchy).

Reply:
I think the ideal system has a little bit of both: hierarchical sharing and hub-and-spoke state access. The injection of this.app down the child-hierarchy is nice to provide some shared capabilities, but it has the downside of linking ALL state, which this changeset will improve.